### ISSUE TITLE
Produce signed .NET container images

### ISSUE THREAD

**Issue by mthalman** [dotnet/dotnet-docker#4589]
Signing a container image involves using a digital signature that includes information about the image, such as its hash value, and is created using a private key. The consumer can then verify this signature using a public key from a trusted source.

This provides the following benefits:

1. Authenticity: Ensures that the image comes from a trusted source and has not been tampered with.
1. Integrity: Guarantees that the image has not been modified since it was signed.
1. Trust: Builds trust between the image publisher and the consumers of the image.

We should produce .NET container images that are signed to provide these additional benefits.

Validation would be supported via the [notation CLI](https://github.com/notaryproject/notation) from the [Notary project](https://notaryproject.dev/docs/overview/). The entire workflow of signing and validation would be the same as is described in [Announcing Image Signing for Windows Containers
](https://techcommunity.microsoft.com/t5/containers/announcing-image-signing-for-windows-containers/ba-p/3792097)

**Comment by mthalman** [comment]
Here's the link for how to configure the infrastructure to produce signed images: https://eng.ms/docs/more/containers-secure-supply-chain/signing (internal link).

**Comment by richlander** [comment]
One option is:

- Start signing with .NET 8 only.
- Prove that out and ensure no issues.
- Wait for folks to ask for signing with older versions.

**Comment by mthalman** [comment]
Related work item: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2060356 (internal link)

**Comment by lbussell** [comment]
[Triage] This will use the ORAS tool to sign the images. An issue should be logged to add the tool to the ImageBuilder Dockerfile. This is also needed for https://github.com/dotnet/docker-tools/issues/1201. 

The signing process doesn't have to happen at build time, so we don't need a solution for using the ORAS tool on Windows agents. Including it in the Linux ImageBuilder image will be sufficient.

**Comment by lbussell** [comment]
A couple of key points to highlight:

- *Images* themselves are not signed. What is signed is a JSON payload containing info about an image digest (which refers to a manifest). Since a digest necessarily changes whenever the content of an image is different, signing the digest of the image is sufficient to attest that we created the image.
- The spec for the signature payload is [HERE](https://github.com/notaryproject/specifications/blob/main/specs/signature-specification.md).

Here's a summary of the mechanics for signing images:

1. Build images as normal
2. For each digest to be signed, create a payload according the the Notary V2 spec:
    ```
    {
      "targetArtifact": {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "digest": "sha256:<...>",
        "size": <length of the digest's manifest text>
      }
    }
    ```
3. Send the payload to our signing service to be signed. Applicable keys for signatures and attestations are listed in [Signing container images (eng.ms)](https://eng.ms/docs/more/containers-secure-supply-chain/signing#step-3:-sign-the-payload-using-esrp-signing-service).
    - This can be accomplished with an Azure DevOps task. There is dnceng-specific documentation for this in ["Converting-AzDo-EsrpCodeSigning-Yml-Task..." (dev.azure.com)](https://dev.azure.com/dnceng/internal/_wiki/wikis/DNCEng%20Services%20Wiki/1240/Converting-AzDo-EsrpCodeSigning-Yml-Task-From-Version-4-to-Version-5-For-Existing-DncEng-Supported-Signing-Operations). 
    This could make our builds look very messy if we need to run a separate pipeline step for each image that was built. If there's no way to sign multiple artifacts with one pipeline step, then I would consider that a point in favor of making image signing a separate pipeline stage.
    - The alternative is to use "ESRP.exe" which appears to be Windows-specific - not ideal for our case.
    - There is also an in-development Az CLI extension that could be used, but it's not well documented and that option needs more investigation.
5. Use ORAS to attach and push the signature to the ACR:
    ```
      oras attach myregistry.azurecr.io/namespace/namespace/hello-world:v0.0.1 \
        --artifact-type 'application/vnd.cncf.notary.signature' \
        ./payload.json:application/cose \
        -a "io.cncf.notary.x509chain.thumbprint#S256=[\"aaaaaaaaaa\", \"bbbbbbbbbb\"]
    ```
    The thumbprint is the list of SHA-256 thumbprints of the signing certificate and certificate chain (including root) that were used for signature generation.

The next step is to run an experiment with signing some of our images in our internal/testing ACR.

**Comment by richlander** [comment]
Would be good if we provided a mini tutorial on how to use the tool for our images. Would also be good to show to how to validate the contents. That would be a step harder but I think worthwhile.

**Comment by lbussell** [comment]
> Would be good if we provided a mini tutorial on how to use the tool for our images. Would also be good to show to how to validate the contents. That would be a step harder but I think worthwhile.

@richlander, I completely agree, I wouldn't consider this feature complete without documentation showing how to verify the signature of an image.

**Comment by mthalman** [comment]
Ideally that documentation would be hosted on the MAR portal since it is not specific to .NET images. Perhaps the portal should also have some indication of which images are signed. /cc @terencet-dev

**Comment by terencet-dev** [comment]
Thanks for the tag and feedback. Support for referrers (such as signatures) is still part of our backlog at the moment. /cc @nhu1997

**Comment by la35-esimplicity** [comment]
Hi there! Are there still plans to sign .NET 8 container images (e.g., `mcr.microsoft.com/dotnet/sdk:8.0*` and `mcr.microsoft.com/dotnet/runtime:*`)? Also, is it 100% confirmed that Docker Content Trust will not be supported for verification (https://github.com/dotnet/dotnet-docker/issues/645)? Thank you!

**Comment by lbussell** [comment]
Hello @la35-esimplicity. At the moment we plan to work on this in the .NET 10 timeframe. When we have the capability to sign images, we'll sign all of our currently supported images. According to our current plans/timeline, that will include .NET 8. This issue will stay up-to-date with the latest progress.

The current plan is to follow the [Notary ("v2") specification](https://github.com/notaryproject/specifications/blob/v1.1.0/specs/signature-specification.md) which is essentially the newer version of Docker Content Trust. Looking at #645 again, my previous comment https://github.com/dotnet/dotnet-docker/issues/645#issuecomment-1781708165 doesn't seem correct. You should be able to set up docker content trust with images signed via the notary project - e.g. https://www.cncf.io/blog/2021/07/28/enforcing-image-trust-on-docker-containers-using-notary/

**Comment by lbussell** [comment]
Signature artifacts are now published for all supported .NET container images. Please [file an issue](https://github.com/dotnet/dotnet-docker/issues/new/choose) if you encounter problems consuming/verifying .NET image signatures. Related links:

- Announcement: https://github.com/dotnet/dotnet-docker/discussions/7150
- Documentation: https://github.com/dotnet/dotnet-docker/blob/main/documentation/image-signatures.md
